Firestore/core/src/firebase/firestore/util/string_format.cc

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+namespace internal {
+
+static const char* kMissing = "<missing>";
+static const char* kInvalid = "<invalid>";
+
+std::string StringFormatPieces(
+    const char* format, std::initializer_list<absl::string_view> pieces) {
+  std::string result;
+
+  const char* format_iter = format;
+  const char* format_end = format + strlen(format);
+  auto pieces_iter = pieces.begin();
+  auto pieces_end = pieces.end();
+
+  while (true) {
```

Comment:
Optional nit: I don't really like `for(ever)` loops, because it's necessary to read the whole implementation to find out how the loop iterates. For this reason, I'll try to provide an alternative version, but as usual, it is entirely optional, feel free to ignore:
```cpp
auto end = format + strlen(format);
const auto next_specifier = [&](const char* i) { return std::find(i, end, '%'); };
for (auto from = format, to = next_specifier(from); to != end;
    from = to + 2, to = next_specifier(from)) {
  result.append(from, to);
  bool complete_specifier = append_specifier(to);
  if (!complete_specifier)
    break;
}
result.append(to, end); // This is now outside the loop
// ...
const auto append_specifier = [&](const char* i) {
  if (i == end) return true;
  if (i + 1 == end) {
    result += '%';
    return false;
  }
  switch (i[1]) {
  // ...
  }
  return true;
}
```

Reply:
I know you're not a fan, but in cases like these I find the contortions required to fit legibly into a traditional `for` loop are a cure worse than the disease. I've pulled some of the details out of the loop.

Reply:
I think the updated version is a lot more readable, thanks for doing this.